http-client/src/main/java/com/facebook/airlift/http/client/thrift/ThriftResponseHandler.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.airlift.http.client.thrift;
+
+import com.facebook.airlift.http.client.HeaderName;
+import com.facebook.airlift.http.client.Request;
+import com.facebook.airlift.http.client.Response;
+import com.facebook.airlift.http.client.ResponseHandler;
+import com.facebook.airlift.http.client.ResponseHandlerUtils;
+import com.facebook.drift.codec.ThriftCodec;
+import com.facebook.drift.transport.netty.codec.Protocol;
+import com.google.common.collect.ImmutableListMultimap;
+import com.google.common.collect.ListMultimap;
+import com.google.common.net.MediaType;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.List;
+
+import static com.google.common.net.HttpHeaders.CONTENT_TYPE;
+import static java.util.Objects.requireNonNull;
+
+public class ThriftResponseHandler<T>
+        implements ResponseHandler<ThriftResponseHandler.ThriftResponse, RuntimeException>
+{
+    private final ThriftCodec<T> thriftCodec;
+
+    public ThriftResponseHandler(ThriftCodec<T> thriftCodec)
+    {
+        this.thriftCodec = requireNonNull(thriftCodec, "thriftCodec is null");
+    }
+
+    @Override
+    public ThriftResponse handleException(Request request, Exception exception)
+    {
+        throw ResponseHandlerUtils.propagate(request, exception);
+    }
+
+    @Override
+    public ThriftResponse handle(Request request, Response response)
+    {
+        T value = null;
+        IllegalArgumentException exception = null;
+        try {
+            Protocol protocol = getThriftProtocol(response.getHeaders());
+            value = ThriftProtocolUtils.read(thriftCodec, protocol, response.getInputStream());
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
```

Comment:
Why special case this for IOException--what is the effect that the protocol throws an IO Exception?

Reply:
It is handled separately for all the other response handlers. From the code we may conclude that, IOException needs to be propagated.  There is something wrong on the server communication so the higher layer should handle this case.

Reply:
Will the higher level need to know to catch the `UncheckedIOException`, or will it be unwrapped automatically?